### PR TITLE
fix: handle panic on null values

### PR DESCRIPTION
**Pull request by Valkyrie00**
Issue related: https://github.com/spotify/confidence-sdk-go/issues/61

This pull request includes several changes to the `pkg/confidence` package, primarily focusing on adding new test cases for resolving null values and handling type mismatches more gracefully. The most important changes are listed below:

### New Test Cases for Null Values:
* Added `TestResolveBoolNullValue`, `TestResolveIntNullValue`, `TestResolveDoubleNullValue`, and `TestResolveStringValueNullValue` to test the resolution of null values for boolean, integer, double, and string flags respectively. (`pkg/confidence/confidence_internal_test.go`, [[1]](diffhunk://#diff-0397b80202ef4a0ca4601c193a901bc7df3c4b81d555c776341f83d73e289c85R38-R50) [[2]](diffhunk://#diff-0397b80202ef4a0ca4601c193a901bc7df3c4b81d555c776341f83d73e289c85R60-R72) [[3]](diffhunk://#diff-0397b80202ef4a0ca4601c193a901bc7df3c4b81d555c776341f83d73e289c85R82-R94) [[4]](diffhunk://#diff-0397b80202ef4a0ca4601c193a901bc7df3c4b81d555c776341f83d73e289c85R104-R116)
* Added nested test cases for null values, such as `TestResolveBoolNestedNullValue`, `TestResolveStringNestedNullValue`, `TestResolveDoubleNestedNullValue`, and `TestResolveIntNestedNullValue`. (`pkg/confidence/confidence_internal_test.go`, [pkg/confidence/confidence_internal_test.goL74-R226](diffhunk://#diff-0397b80202ef4a0ca4601c193a901bc7df3c4b81d555c776341f83d73e289c85L74-R226))

### Template Response Enhancements:
* Updated `templateResponseWithFlagName` to include null values for various keys, enhancing the template response to handle null values in different contexts. (`pkg/confidence/confidence_internal_test.go`, [[1]](diffhunk://#diff-0397b80202ef4a0ca4601c193a901bc7df3c4b81d555c776341f83d73e289c85R324-R343) [[2]](diffhunk://#diff-0397b80202ef4a0ca4601c193a901bc7df3c4b81d555c776341f83d73e289c85R353-R381) [[3]](diffhunk://#diff-0397b80202ef4a0ca4601c193a901bc7df3c4b81d555c776341f83d73e289c85R392-R417)

### Utility Function Improvements:
* Modified `replaceNumbers` to handle `json.Number` more robustly, ensuring proper conversion to `float64` and `int64` even when values are null. (`pkg/confidence/utils.go`, [pkg/confidence/utils.goL163-R183](diffhunk://#diff-06cb179db5dad83bda71a638eade2b8b1af1777f374defd1263ef764fd38cf3aL163-R183))
* Enhanced type conversion functions (`ToBoolResolutionDetail`, `ToStringResolutionDetail`, `ToFloatResolutionDetail`, `ToIntResolutionDetail`, and `ToObjectResolutionDetail`) to check for null values before attempting type conversion. (`pkg/confidence/utils.go`, [[1]](diffhunk://#diff-06cb179db5dad83bda71a638eade2b8b1af1777f374defd1263ef764fd38cf3aL212-R220) [[2]](diffhunk://#diff-06cb179db5dad83bda71a638eade2b8b1af1777f374defd1263ef764fd38cf3aL241-R249) [[3]](diffhunk://#diff-06cb179db5dad83bda71a638eade2b8b1af1777f374defd1263ef764fd38cf3aL272-R280) [[4]](diffhunk://#diff-06cb179db5dad83bda71a638eade2b8b1af1777f374defd1263ef764fd38cf3aL311-R319) [[5]](diffhunk://#diff-06cb179db5dad83bda71a638eade2b8b1af1777f374defd1263ef764fd38cf3aL333-R341)

### Error Message Corrections:
* Corrected error messages in type conversion functions to accurately reflect the type being converted, e.g., changing "Unable to convert response property to boolean" to "Unable to convert response property to string". (`pkg/confidence/utils.go`, [[1]](diffhunk://#diff-06cb179db5dad83bda71a638eade2b8b1af1777f374defd1263ef764fd38cf3aL250-R258) [[2]](diffhunk://#diff-fcc0dacfa6a0f9abda4e7300b89ff63046d2816efba47d2cda42d8f4c0c95cd2L484-R484)

### CODE REVIEW

File: pkg/confidence/confidence_internal_test.go [hunk 1]
```diff
@@ -44,6 +44,15 @@ func TestResolveIntValue(t *testing.T) {
 	assert.Equal(t, int64(40), evalDetails.Value)
 }
 
+func TestResolveIntNullValue(t *testing.T) {
+	client := client(t, templateResponse(), nil)
+	client.PutContext("targeting_key", "user1")
+
+	evalDetails := client.GetIntFlag(context.Background(), "test-flag.integer-key-null", 99)
+
+	assert.Equal(t, int64(99), evalDetails.Value)
```
Comment (fabriziodemaria) [inline pkg/confidence/confidence_internal_test.go:53]: We should add the full coverage of `evalDetails`. The expected results for this specific evaluation should be [this](https://github.com/spotify/confidence-sdk-go/commit/4bcfcaf2aadb82c054b793a4943d5708811bf41b), while in the current state an ErrorCode is reported

**Review by fabriziodemaria — CHANGES_REQUESTED** [review]
Thank you for the contribution! 

It's a great start to fix the problem with `null`s in the payload, however I noticed some remaining issues in how `evalDetails` is generated in this case (left a comment about it).

Overall, I would like to improve the handling of `null` values for all types in the same PR (i.e. bools, strings, etc.), thus avoiding staggered fix releases that might be even more confusing for the user of the library.

We understand if you decide to not move forward with this, and in that case we will eventually pick up from where you started and complete the fix

File: pkg/confidence/utils.go [hunk 2]
```diff
@@ -247,7 +255,7 @@ func ToStringResolutionDetail(res InterfaceResolutionDetail,
 			}
 		}
 
-		err := NewTypeMismatchResolutionError("Unable to convert response property to boolean")
+		err := NewTypeMismatchResolutionError("Unable to convert response property to string")
```
Comment (Valkyrie00) [inline pkg/confidence/utils.go:258]: @fabriziodemaria I also found this little typo

File: pkg/confidence/utils.go [hunk 3]
```diff
@@ -308,7 +316,7 @@ func ToObjectResolutionDetail(res InterfaceResolutionDetail, defaultValue interf
 			}
 		}
 
-		err := NewTypeMismatchResolutionError("Unable to convert response property to float")
+		err := NewTypeMismatchResolutionError("Unable to convert response property to object")
```
Comment (Valkyrie00) [inline pkg/confidence/utils.go:319]: @fabriziodemaria also here, but in generally about this part, I did not understand the true nature of ToObjectResolutionDetail. I'm probably missing some internal decision thinking, including about the package provider.

File: pkg/confidence/confidence_internal_test.go (hunk 1 above)
Comment (Valkyrie00) [reply pkg/confidence/confidence_internal_test.go:66]: Oww okok I see what you mean, you are absolutely right 👍

Comment (Valkyrie00) [reply pkg/confidence/confidence_internal_test.go:66]: I've added this control here and in the other tests

**Comment by Valkyrie00** [comment]
> Thank you for the contribution!
> 
> This is a great start to fix the problem with `null` values in the payload. However, I noticed some remaining issues with how `evalDetails` is generated in this case (left a comment about it).
> 
> Overall, I would like to improve the handling of `null` values for all types (booleans, strings, etc.) in the same PR, thus avoiding staggered fix releases which might cause even more confusion for the user of the library.
> 
> We understand if you decide to not move forward with this, and in that case we will eventually pick up from where you started and complete the fix ourselves.

Really don't worry about it, I hope this helps 😃 I make these contributions because I'm happy to try new things.

I've made the changes to support the various types, and all the tests should be there as well.
In general, the main additional change was leveraging the different `ResolutionDetail` instances to ensure that if the only return reason is `TargetingMatchReason` and its value is not `nil`, then it performs the cast of the value.
If that's not the case, it falls back to the default return value, which was already the existing behavior.

I hope I haven't forgotten anything, or in case you don't like the approach/implementation let me know I can study other ways.

File: pkg/confidence/confidence_internal_test.go (hunk 1 above)
Comment (fabriziodemaria) [reply pkg/confidence/confidence_internal_test.go:66]: Great!

File: pkg/confidence/utils.go (hunk 3 above)
Comment (fabriziodemaria) [reply pkg/confidence/utils.go:319]: The idea of `ToObjectResolutionDetail` is to be able to return the entire flag, rather than accessing a specific property of that flag. I think we still miss a strong specification regarding this API:

- What should happen when some/all the properties of the flag gotten from the backend are `null`? Should we replace only those null properties with the corresponding property values from the the defaultValue passed in?
- What should happen if the schema of the flag from the backend and the schema of the defaultValue passed in are different?

This problem is separate from the scope of the PR, so I suggest we go ahead and start merging these improvements!

**Review by fabriziodemaria — APPROVED** [review]
LGTM, and great improvements! Thank you again for your contributions
